File: frontend/shared/protocol.ts
// discodb2 — CANONICAL protocol & data-model types (frontend/shared).
//
// SOURCE OF TRUTH: docs/DESIGN.md §3 (THE CONTRACT) and §3.5 (data model).
// This file is framework-free (no Svelte/Vite/DOM-only deps) so it can be
// imported by cockpit, copilot, a Web Worker, or a plain Node test runner.
//
// The cockpit/copilot builder agents implement the protocol independently;
// this module is the version everything CONSOLIDATES onto later. When you
// change the wire format here, you are changing the contract — bump VERSION
// and note the breaking change in DESIGN.md.
//
// Endianness: little-endian everywhere (DESIGN §3.2). Classic CAN only (DLC
// 0..8); CAN-FD is a future v2.

/* ────────────────────────────────────────────────────────────────────────
 * §3.2 — Binary frame stream (batch header + fixed 20-byte records)
 * ──────────────────────────────────────────────────────────────────────── */

/** Wire protocol version. Lives in byte 0 of every batch header. */
export const PROTOCOL_VERSION = 1 as const;

/** Batch header is a fixed 12 bytes. */
export const BATCH_HEADER_BYTES = 12 as const;

/** Each CAN record is a fixed 20 bytes. */
export const RECORD_BYTES = 20 as const;

/** Max classic-CAN payload length. */
export const MAX_DLC = 8 as const;

/** Batch header flag bits (byte 1 of the header). */
export const BatchFlag = {
  /** bit0: this batch was produced by the replay source (not a live bus). */
  REPLAY: 1 << 0,
} as const;
export type BatchFlag = (typeof BatchFlag)[keyof typeof BatchFlag];

/** Per-record flag bits (offset 9 of a record). */
export const RecFlag = {
  /** bit0: Remote Transmission Request. Other bits reserved (must be 0). */
  RTR: 1 << 0,
} as const;
export type RecFlag = (typeof RecFlag)[keyof typeof RecFlag];

/**
 * can_id bit layout (offset 4, u32, DESIGN §3.2).
 * bits 0–28 = identifier · bit30 = error frame · bit31 = extended (29-bit id).
 * (Mirrors the Linux SocketCAN can_id convention; bit29 RTR is carried in
 * rec_flags here, not in can_id.)
 */
export const CanIdBits = {
  /** Mask for the 29-bit identifier field (bits 0–28). */
  ID_MASK: 0x1fffffff,
  /** bit30: error frame. */
  ERROR: 1 << 30,
  /** bit31: extended (29-bit) identifier. Use >>> 0 to keep it unsigned. */
  EXTENDED: (1 << 31) >>> 0,
} as const;

/** A single decoded CAN record (the JS-side shape of a 20-byte record). */
export interface CanRecord {
  /** Offset from the batch `baseTimeUs`, microseconds (u32). */
  dtUs: number;
  /** 11-bit or 29-bit identifier (already masked to bits 0–28). */
  canId: number;
  /** Data length code, 0..8. */
  dlc: number;
  /** True if the extended (29-bit) flag was set. */
  isExtended: boolean;
  /** True if the error-frame flag was set. */
  isError: boolean;
  /** True if the RTR rec_flag bit was set. */
  isRtr: boolean;
  /** Payload. Exactly `dlc` bytes; the wire pads the rest with zeros. */
  data: Uint8Array;
}

/** A decoded batch: header fields plus its records. */
export interface CanBatch {
  /** Protocol version from header byte 0 (should equal PROTOCOL_VERSION). */
  version: number;
  /** True if BatchFlag.REPLAY was set. */
  replay: boolean;
  /** Monotonic microsecond base time for the batch (u64, header). */
  baseTimeUs: bigint;
  /** Decoded records (length === header `count`). */
  records: CanRecord[];
}

/**
 * Encode a batch to the §3.2 binary wire format.
 *
 * Pure: returns a fresh ArrayBuffer, mutates nothing. Throws RangeError on
 * out-of-spec input (DLC > 8, count > u16, etc.) so encode bugs fail loud
 * instead of producing a malformed batch on the hot path.
 */
export function encodeBatch(batch: {
  baseTimeUs: bigint;
  replay?: boolean;
  records: ReadonlyArray<{
    dtUs: number;
    canId: number;
    dlc: number;
    isExtended?: boolean;
    isError?: boolean;
    isRtr?: boolean;
    data: Uint8Array | ReadonlyArray<number>;
  }>;
}): ArrayBuffer {
  const count = batch.records.length;
  if (count > 0xffff) {
    throw new RangeError(`batch count ${count} exceeds u16 max (65535)`);
  }

  const buf = new ArrayBuffer(BATCH_HEADER_BYTES + count * RECORD_BYTES);
  const view = new DataView(buf);
  const u8 = new Uint8Array(buf);

  // Header (12 bytes, little-endian).
  view.setUint8(0, PROTOCOL_VERSION);
  view.setUint8(1, batch.replay ? BatchFlag.REPLAY : 0);
  view.setUint16(2, count, true);
  view.setBigUint64(4, batch.baseTimeUs, true);

  // Records (20 bytes each).
  for (let i = 0; i < count; i++) {
    const r = batch.records[i];
    const off = BATCH_HEADER_BYTES + i * RECORD_BYTES;

    if (r.dlc < 0 || r.dlc > MAX_DLC) {
      throw new RangeError(`record ${i}: dlc ${r.dlc} out of range 0..8`);
    }
    const idField = r.canId & CanIdBits.ID_MASK;
    if ((r.canId & ~CanIdBits.ID_MASK) !== 0 && r.canId <= CanIdBits.ID_MASK) {
      // canId carried stray high bits but is itself a small number: caller
      // likely OR-ed flag bits into canId. Reject — flags are explicit fields.
      throw new RangeError(`record ${i}: canId has bits above 0..28 set; pass flags via isExtended/isError`);
    }

    let canIdWord = idField >>> 0;
    if (r.isExtended) canIdWord = (canIdWord | CanIdBits.EXTENDED) >>> 0;
    if (r.isError) canIdWord = (canIdWord | CanIdBits.ERROR) >>> 0;

    let recFlags = 0;
    if (r.isRtr) recFlags |= RecFlag.RTR;

    view.setUint32(off + 0, r.dtUs >>> 0, true);
    view.setUint32(off + 4, canIdWord, true);
    view.setUint8(off + 8, r.dlc);
    view.setUint8(off + 9, recFlags);
    view.setUint16(off + 10, 0, true); // reserved
    // data[8]: copy dlc bytes, the ArrayBuffer is already zero-filled past dlc.
    for (let b = 0; b < r.dlc; b++) {
      u8[off + 12 + b] = r.data[b] & 0xff;
    }
  }

  return buf;
}

/**
 * Decode a §3.2 binary batch.
 *
 * Pure: reads `buf`, allocates fresh output, mutates nothing. The returned
 * record `data` arrays are views' copies (length === dlc). Throws RangeError
 * if the buffer is too short for its declared `count` — a truncated frame is
 * a protocol error, not silently tolerated.
 *
 * `buf` may be an ArrayBuffer or any ArrayBufferView (e.g. a Uint8Array
 * received from a WebSocket); the view's byteOffset/byteLength are honored.
 */
export function decodeBatch(buf: ArrayBuffer | ArrayBufferView): CanBatch {
  let ab: ArrayBuffer;
  let byteOffset = 0;
  let byteLength: number;
  if (ArrayBuffer.isView(buf)) {
    ab = buf.buffer as ArrayBuffer;
    byteOffset = buf.byteOffset;
    byteLength = buf.byteLength;
  } else {
    ab = buf;
    byteLength = buf.byteLength;
  }

  if (byteLength < BATCH_HEADER_BYTES) {
    throw new RangeError(`batch shorter than ${BATCH_HEADER_BYTES}-byte header (got ${byteLength})`);
  }

  const view = new DataView(ab, byteOffset, byteLength);
  const version = view.getUint8(0);
  const flags = view.getUint8(1);
  const count = view.getUint16(2, true);
  const baseTimeUs = view.getBigUint64(4, true);

  const need = BATCH_HEADER_BYTES + count * RECORD_BYTES;
  if (byteLength < need) {
    throw new RangeError(`batch truncated: count=${count} needs ${need} bytes, got ${byteLength}`);
  }

  const records: CanRecord[] = new Array(count);
  for (let i = 0; i < count; i++) {
    const off = BATCH_HEADER_BYTES + i * RECORD_BYTES;
    const dtUs = view.getUint32(off + 0, true);
    const canIdWord = view.getUint32(off + 4, true);
    const dlcRaw = view.getUint8(off + 8);
    const recFlags = view.getUint8(off + 9);
    // off+10..11 reserved (ignored on decode).

    const dlc = Math.min(dlcRaw, MAX_DLC);
    const data = new Uint8Array(dlc);
    for (let b = 0; b < dlc; b++) {
      data[b] = view.getUint8(off + 12 + b);
    }

    records[i] = {
      dtUs,
      canId: canIdWord & CanIdBits.ID_MASK,
      dlc,
      isExtended: (canIdWord & CanIdBits.EXTENDED) !== 0,
      isError: (canIdWord & CanIdBits.ERROR) !== 0,
      isRtr: (recFlags & RecFlag.RTR) !== 0,
      data,
    };
  }

  return {
    version,
    replay: (flags & BatchFlag.REPLAY) !== 0,
    baseTimeUs,
    records,
  };
}

/** Absolute monotonic timestamp of a record = batch base + record offset. */
export function recordTimeUs(batch: CanBatch, rec: CanRecord): bigint {
  return batch.baseTimeUs + BigInt(rec.dtUs);
}

/* ────────────────────────────────────────────────────────────────────────
 * §3.3 — Control messages (client → server, JSON text)
 * ──────────────────────────────────────────────────────────────────────── */

/** Which frontend is connecting. */
export type ClientKind = "cockpit" | "copilot";

/** CAN source the backend can open (DESIGN §3.3 / §5). */
export type CanSource = "sim" | "socketcan" | "gs_usb" | "slcan" | "replay";

export interface HelloMsg {
  type: "hello";
  client: ClientKind;
}

export interface StartMsg {
  type: "start";
  source: CanSource;
  bitrate: number;
  /**
   * Listen-only request. The SERVER enforces this for any live source
   * (DESIGN §4.1): a request to disable it is refused or clamped. This field
   * is advisory from the client's side — the guarantee is server-side.
   */
  listen_only: boolean;
  /** Replay only: file to stream. */
  file?: string;
}

export interface StopMsg {
  type: "stop";
}

export interface RecordStartMsg {
  type: "record_start";
  /** Optional human name for the recording. */
  name?: string;
}

export interface RecordStopMsg {
  type: "record_stop";
}

export interface ListFilesMsg {
  type: "list_files";
}

/** Discriminated union of every client→server control message. */
export type ClientMessage =
  | HelloMsg
  | StartMsg
  | StopMsg
  | RecordStartMsg
  | RecordStopMsg
  | ListFilesMsg;

/** Serialize a control message to a JSON text frame. */
export function encodeControl(msg: ClientMessage): string {
  return JSON.stringify(msg);
}

/**
 * Parse a JSON text frame into a control message.
 *
 * Returns the parsed object typed as ClientMessage if it has a known `type`,
 * otherwise null. Kept intentionally tiny: the backend is the real validator
 * (DESIGN §4 — the guarantee never depends on the client).
 */
export function decodeControl(text: string): ClientMessage | null {
  let obj: unknown;
  try {
    obj = JSON.parse(text);
  } catch {
    return null;
  }
  if (typeof obj !== "object" || obj === null) return null;
  const type = (obj as { type?: unknown }).type;
  switch (type) {
    case "hello":
    case "start":
    case "stop":
    case "record_start":
    case "record_stop":
    case "list_files":
      return obj as ClientMessage;
    default:
      return null;
  }
}

/* ────────────────────────────────────────────────────────────────────────
 * §3.4 — Status / Health (server → client text, and GET /health JSON)
 * ──────────────────────────────────────────────────────────────────────── */

/** Bus run state. "LIVE" for a real/sim bus actively streaming. */
export type BusState = "LIVE" | "STOPPED" | "ERROR" | "REPLAY";

export interface BusHealth {
  bitrate: number;
  state: BusState;
  fps: number;
  fps_avg: number;
  total: number;
  unique_ids: number;
  errors: number;
  /** Milliseconds since the last frame (relative; never wall clock). */
  last_frame_ms: number;
  /** Estimated bus load, 0.0..1.0. */
  bus_load: number;
}

export interface StreamHealth {
  clients: number;
  out_bps: number;
  dropped: number;
}

export interface RecordHealth {
  active: boolean;
  file: string | null;
  size: number;
  disk_free: number;
}

export interface ProcHealth {
  cpu: number;
  rss: number;
  reader_q: number;
  ws_q: number;
}

/** The full health/status object (DESIGN §3.4). */
export interface Health {
  /** Seconds since backend start (monotonic; the Pi has no RTC). */
  uptime_s: number;
  source: CanSource;
  listen_only: boolean;
  /** Recording name if active, else null. */
  recording: string | null;
  bus: BusHealth;
  stream: StreamHealth;
  record: RecordHealth;
  proc: ProcHealth;
}

/** server→client: directory listing in response to list_files. */
export interface FilesMsg {
  type: "files";
  files: string[];
}

/** server→client: error notification. */
export interface ErrorMsg {
  type: "error";
  message: string;
}

/**
 * Server→client text frames carry either a tagged message (`files`/`error`)
 * or a bare Health object (which has no `type` field, per §3.4).
 */
export type ServerMessage = FilesMsg | ErrorMsg | Health;

/** True if a parsed server text frame is a Health snapshot (no `type` tag). */
export function isHealth(msg: ServerMessage): msg is Health {
  return !("type" in msg) && "uptime_s" in msg;
}

/* ────────────────────────────────────────────────────────────────────────
 * §3.5 — Data model (Signal / FrameDef / Project)
 * ──────────────────────────────────────────────────────────────────────── */

/** Byte order of a signal's bit range within its frame. */
export type ByteOrder = "big" | "little";

/** One signal: a bit range within a frame mapped to a physical value. */
export interface Signal {
  id: string;
  frameId: number;
  isExtended: boolean;
  bitStart: number;
  bitLength: number;
  byteOrder: ByteOrder;
  /** Physical = raw * factor + offset. */
  factor: number;
  offset: number;
  unit: string;
  name: string;
}

/** A CAN frame definition: an arbitration id carrying multiple signals. */
export interface FrameDef {
  id: number;
  isExtended: boolean;
  name: string;
  signals: Signal[];
}

/** A reverse-engineering project: a named set of frame definitions. */
export interface Project {
  name: string;
  frames: FrameDef[];
}

File: backend/discodb2_backend/protocol.py
"""Binary frame-stream wire protocol (DESIGN.md §3.2).

This module is the single source of truth for the on-the-wire byte layout of the
batched CAN frame stream. It deliberately depends on nothing but ``struct`` so it
can be imported by tests and by frontends' reference implementations alike, and
so it stays trivially portable to the ARMv6 Pi target (no numpy).

Wire layout (little-endian, batched ~20-50 ms):

    Batch header (12 bytes):
        off 0  version    u8   (== PROTOCOL_VERSION, currently 1)
        off 1  flags      u8   bit0: 1 = replay batch (else live)
        off 2  count      u16  number of records that follow
        off 4  base_t_us  u64  monotonic microseconds, batch time base

    Then ``count`` x 20-byte record:
        off 0  dt_us      u32  offset from base_t_us (this frame's monotonic us)
        off 4  can_id     u32  bits 0-28 id, bit30 error, bit31 extended (29-bit)
        off 8  dlc        u8   0..8 (classic CAN only)
        off 9  rec_flags  u8   bit0 RTR (other bits reserved, 0)
        off 10 reserved   u16  0
        off 12 data       u8[8] bytes with index >= dlc are 0

Classic CAN only (dlc <= 8). CAN-FD is a future v2.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field

PROTOCOL_VERSION = 1

# --- batch header ---------------------------------------------------------
HEADER_FORMAT = "<BBHQ"  # version u8, flags u8, count u16, base_t_us u64
HEADER_SIZE = struct.calcsize(HEADER_FORMAT)
assert HEADER_SIZE == 12, HEADER_SIZE

# --- per-frame record -----------------------------------------------------
RECORD_FORMAT = "<IIBBH8s"  # dt_us u32, can_id u32, dlc u8, rec_flags u8, reserved u16, data 8s
RECORD_SIZE = struct.calcsize(RECORD_FORMAT)
assert RECORD_SIZE == 20, RECORD_SIZE

# --- batch flags ----------------------------------------------------------
BATCH_FLAG_REPLAY = 0x01  # bit0 of header flags

# --- can_id bitfield (per §3.2) ------------------------------------------
CAN_ID_MASK = 0x1FFFFFFF  # bits 0-28
CAN_EFF_FLAG = 0x80000000  # bit31: extended (29-bit) id
CAN_ERR_FLAG = 0x40000000  # bit30: error frame
# (bit29 is reserved; intentionally not used. This mirrors Linux SocketCAN's
#  CAN_EFF_FLAG/CAN_ERR_FLAG bit positions, with RTR carried in rec_flags.)

# --- rec_flags bitfield ---------------------------------------------------
REC_FLAG_RTR = 0x01  # bit0: remote transmission request

_HEADER_STRUCT = struct.Struct(HEADER_FORMAT)
_RECORD_STRUCT = struct.Struct(RECORD_FORMAT)


@dataclass(slots=True)
class Frame:
    """A single decoded CAN frame, monotonic-µs timestamped.

    ``can_id`` is the *raw* arbitration id (bits 0-28 only); ``is_extended``,
    ``is_error`` and ``is_rtr`` carry the flag bits separately so callers never
    have to mask. ``t_us`` is the absolute monotonic microsecond timestamp; the
    batcher converts it to a per-batch ``dt_us`` offset on the wire.
    """

    t_us: int
    can_id: int
    dlc: int
    data: bytes = b""
    is_extended: bool = False
    is_error: bool = False
    is_rtr: bool = False

    def __post_init__(self) -> None:
        # Normalise data to exactly the dlc-relevant bytes; the wire format
        # zero-pads to 8. Guard classic-CAN dlc bounds.
        if self.dlc < 0 or self.dlc > 8:
            raise ValueError(f"classic CAN dlc must be 0..8, got {self.dlc}")
        if len(self.data) > 8:
            raise ValueError(f"data longer than 8 bytes: {len(self.data)}")

    def encoded_can_id(self) -> int:
        """The 32-bit can_id field exactly as it goes on the wire."""
        value = self.can_id & CAN_ID_MASK
        if self.is_extended:
            value |= CAN_EFF_FLAG
        if self.is_error:
            value |= CAN_ERR_FLAG
        return value


def encode_record(frame: Frame, base_t_us: int) -> bytes:
    """Pack one :class:`Frame` into its 20-byte wire record."""
    dt_us = frame.t_us - base_t_us
    if dt_us < 0:
        raise ValueError(f"frame t_us {frame.t_us} precedes base {base_t_us}")
    if dt_us > 0xFFFFFFFF:
        raise ValueError(f"dt_us {dt_us} overflows u32; rebatch with a closer base")
    rec_flags = REC_FLAG_RTR if frame.is_rtr else 0
    # Zero-pad data to 8 bytes (struct '8s' truncates/pads, but we pad here so
    # bytes >= dlc are explicitly 0 even if the caller over-supplied).
    data = frame.data[: frame.dlc].ljust(8, b"\x00")
    return _RECORD_STRUCT.pack(
        dt_us,
        frame.encoded_can_id(),
        frame.dlc,
        rec_flags,
        0,  # reserved
        data,
    )


def encode_batch(frames: list[Frame], base_t_us: int, *, replay: bool = False) -> bytes:
    """Pack a list of frames into a single binary batch (header + records).

    ``base_t_us`` is the batch time base; each record stores ``t_us - base_t_us``.
    Pick a base <= the earliest frame's ``t_us`` (the batcher uses the first
    frame's timestamp).
    """
    if len(frames) > 0xFFFF:
        raise ValueError(f"batch of {len(frames)} exceeds u16 count")
    flags = BATCH_FLAG_REPLAY if replay else 0
    header = _HEADER_STRUCT.pack(PROTOCOL_VERSION, flags, len(frames), base_t_us)
    parts = [header]
    parts.extend(encode_record(f, base_t_us) for f in frames)
    return b"".join(parts)


@dataclass(slots=True)
class BatchHeader:
    version: int
    flags: int
    count: int
    base_t_us: int

    @property
    def is_replay(self) -> bool:
        return bool(self.flags & BATCH_FLAG_REPLAY)


@dataclass(slots=True)
class DecodedBatch:
    header: BatchHeader
    frames: list[Frame] = field(default_factory=list)


def decode_record(buf: bytes, base_t_us: int, offset: int = 0) -> Frame:
    """Decode one 20-byte record located at ``offset`` within ``buf``."""
    dt_us, raw_id, dlc, rec_flags, _reserved, data = _RECORD_STRUCT.unpack_from(buf, offset)
    return Frame(
        t_us=base_t_us + dt_us,
        can_id=raw_id & CAN_ID_MASK,
        dlc=dlc,
        data=bytes(data[:dlc]),
        is_extended=bool(raw_id & CAN_EFF_FLAG),
        is_error=bool(raw_id & CAN_ERR_FLAG),
        is_rtr=bool(rec_flags & REC_FLAG_RTR),
    )


def decode_batch(buf: bytes) -> DecodedBatch:
    """Decode a full binary batch back into a :class:`DecodedBatch`.

    The inverse of :func:`encode_batch`; used by tests and reference clients.
    """
    if len(buf) < HEADER_SIZE:
        raise ValueError(f"buffer too short for header: {len(buf)} < {HEADER_SIZE}")
    version, flags, count, base_t_us = _HEADER_STRUCT.unpack_from(buf, 0)
    if version != PROTOCOL_VERSION:
        raise ValueError(f"unsupported protocol version {version}")
    expected = HEADER_SIZE + count * RECORD_SIZE
    if len(buf) != expected:
        raise ValueError(
            f"batch length {len(buf)} != header({HEADER_SIZE}) + "
            f"{count} x record({RECORD_SIZE}) = {expected}"
        )
    header = BatchHeader(version=version, flags=flags, count=count, base_t_us=base_t_us)
    frames = [
        decode_record(buf, base_t_us, HEADER_SIZE + i * RECORD_SIZE) for i in range(count)
    ]
    return DecodedBatch(header=header, frames=frames)
